plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java

```diff
@@ -108,55 +111,62 @@ static AWSCredentialsProvider buildCredentials(Logger logger, Ec2ClientSettings
-
     /**
-     * Refreshes the settings for the AmazonEC2 client. New clients will be build
-     * using these new settings. Old client is usable until released. On release it
+     * Refreshes the settings for the AmazonEC2 client. The new client will be build
+     * using these new settings. The old client is usable until released. On release it
      * will be destroyed instead of being returned to the cache.
      */
     @Override
-    public synchronized Ec2ClientSettings refreshAndClearCache(Ec2ClientSettings clientSettings) {
-        // shutdown all unused clients
-        // others will shutdown on their respective release
-        releaseCachedClient();
-        final Ec2ClientSettings prevSettings = this.clientSettings;
-        this.clientSettings = clientSettings;
-        return prevSettings;
+    public void refreshAndClearCache(Ec2ClientSettings clientSettings) {
+        final LazyAmazonEc2Reference newClient = new LazyAmazonEc2Reference(() -> buildClient(clientSettings));
+        final LazyAmazonEc2Reference oldClient = this.lazyClientReference.getAndSet(newClient);
+        if (oldClient != null) {
+            oldClient.clear();
+        }
     }
 
     @Override
     public void close() {
-        releaseCachedClient();
-    }
-
-    private synchronized void releaseCachedClient() {
-        if (this.clientReference == null) {
-            return;
+        final LazyAmazonEc2Reference clientReference = this.lazyClientReference.getAndSet(null);
+        if (clientReference != null) {
+            clientReference.clear();
         }
-        // the client will shutdown when it will not be used anymore
-        this.clientReference.decRef();
-        // clear the cached client, it will be build lazily
-        this.clientReference = null;
         // shutdown IdleConnectionReaper background thread
         // it will be restarted on new client usage
         IdleConnectionReaper.shutdown();
     }
 
+    private static class LazyAmazonEc2Reference extends Lazy<AmazonEc2Reference, Exception> {
+
+        LazyAmazonEc2Reference(Supplier<AmazonEC2> supplier) {
+            super(() -> {
+                final AmazonEC2 client = Objects.requireNonNull(supplier.get());
+                // wrap supplier result in a reference counted object with initial count set to 1
+                final AmazonEc2Reference clientReference = new AmazonEc2Reference(client);
+                clientReference.incRef();
+                return clientReference;
+            }, clientReference -> clientReference.decRef());
+        }
+
+        @Override
+        public AmazonEc2Reference getOrCompute() {
+            AmazonEc2Reference result;
+            try {
+                result = super.getOrCompute();
+            } catch (final Exception e) {
+                throw new ElasticsearchException("This is unexpected", e);
+            }
+            // the count will be decreased by {@codee AmazonEc2Reference#close}
+            result.incRef();
```

Comment:
we increment reference twice on creation?

Reply:
Yes, this was a mistake, good catch. Ideally this should've been catched by a test. I think there's a TODO (if not for this plugin, for S3) about this test missing. I will address in a follow-up PR.